core/src/test/java/com/datastax/oss/driver/api/core/cql/StatementBuilderTest.java

```diff
@@ -99,4 +127,72 @@ public void should_match_set_routing_key_vararg() {
     builderStmt = builder.setRoutingKey(buff2, buff1).build();
     assertThat(expectedStmt.getRoutingKey()).isNotEqualTo(builderStmt.getRoutingKey());
   }
+
+  @Test
+  public void should_correctly_set_routing_key_on_boundstatement() {
+    PreparedStatement preparedStatement = mockPreparedStatement("UPDATE foo SET v=? WHERE k=?",
+            ImmutableMap.of("v", DataTypes.INT, "k", DataTypes.INT));
+    BoundStatement boundStatement = preparedStatement.boundStatementBuilder().build();
+    // we should not have a routing key set
+    assertThat(boundStatement.getRoutingKey()).isNull();
+    boundStatement = boundStatement.set(0, 1, Integer.class).set(1, 2, Integer.class);
+    // we should have a non-empty routing key now
+    assertThat(boundStatement.getRoutingKey()).isNotNull();
+    assertThat(boundStatement.getRoutingKey().hasRemaining()).isTrue();
+  }
+
+  @Test
+  public void should_correctly_set_routing_key_on_boundstatementbuilder() {
+    PreparedStatement preparedStatement = mockPreparedStatement("UPDATE foo SET v=? WHERE k=?",
+            ImmutableMap.of("v", DataTypes.INT, "k", DataTypes.INT));
+    BoundStatement boundStatement = preparedStatement.boundStatementBuilder().build();
+    BoundStatement copy = new BoundStatementBuilder(boundStatement).build();
+    // we should not have a routing key set
+    assertThat(copy.getRoutingKey()).isNull();
+    copy = copy.set(0, 1, Integer.class).set(1, 2, Integer.class);
+    // we should have a non-empty routing key now
+    assertThat(copy.getRoutingKey()).isNotNull();
+    assertThat(copy.getRoutingKey().hasRemaining()).isTrue();
+  }
+
+  // copied from RequestLogFormatterTest, we should move somewhere to share b/w tests
```

Comment:
hacked the test together by copying this code from `RequestLogFormatterTest`, I'm looking to you folks for advice on where this code should actually live

Reply:
This functionality is probably something that's more easily validated in an integration test.  PreparedStatementIT already does some validation of the ability to compute routing information when the partition key is bound; a test for this functionality should slot in next to it pretty nicely.

I was doing some testing with something like the following:

```java
  @Test
  public void should_return_null_routing_information_when_single_partition_key_is_unbound() {
    should_return_null_routing_information_when_single_partition_key_is_unbound(
            "SELECT a FROM prepared_statement_test WHERE a = ?");
    should_return_null_routing_information_when_single_partition_key_is_unbound(
            "INSERT INTO prepared_statement_test (a) VALUES (?)");
    should_return_null_routing_information_when_single_partition_key_is_unbound(
            "UPDATE prepared_statement_test SET b = 1 WHERE a = ?");
    should_return_null_routing_information_when_single_partition_key_is_unbound(
            "DELETE FROM prepared_statement_test WHERE a = ?");
  }

  private void should_return_null_routing_information_when_single_partition_key_is_unbound(String queryString) {

    CqlSession session = sessionRule.session();
    BoundStatement boundStatement = session.prepare(queryString).bind();
    assertThat(boundStatement.getRoutingKey()).isNull();
  }
```

This test fails with the existing 4.x code but passes with the changes referenced above.

Reply:
thank you, much cleaner than my hacked-together test!

Reply:
And for simplicity I'm going to go with exactly what you have here and not worry about the extra copying tests